Adding InterApp Table Clean up

BautiSLC opened the pull request.

BautiSLC committed:
Table Cleanup logic

BautiSLC committed:
Add and Remove Repos is now included in the interapp table

BlakeSLC commented on protocol.xml:
```diff
@@ -5250,6 +5250,64 @@ Are you sure you want to continue.]]>
+				<Range>
+					<Low>300</Low>
+				</Range>
+				<Steps>60</Steps>
+				<Positions>
```
can be made into options="time:minute" instead of steps

BlakeSLC reviewed (approved):
minor remark, hopefully Arne can also give us his feedback

ArneMaes0 commented on protocol.xml:
```diff
@@ -5432,6 +5491,27 @@ Are you sure you want to continue.]]>
+			<Type>read</Type>
+			<Information>
+				<Subtext>Time the InterApp message was received.</Subtext>
+			</Information>
```
Perhaps renaming this one to Received At instead of Request Time as I feel that more clearly describes what this parameter is.

ArneMaes0 commented on protocol.xml:
```diff
@@ -5452,6 +5532,7 @@ Are you sure you want to continue.]]>
 		<QAction id="20002" name="HTTP Poll Manager Updates" encoding="csharp" row="true" triggers="21006;21053" />
 
 		<QAction id="9000000" name="Process Incoming InterApp Messages" encoding="csharp" triggers="9000000"/>
+		<QAction id="9000049" name="IAC Messages Cleanup" encoding="csharp" triggers="9000049"/>
```
There is already a Table Cleanup row in the Poll Manager you can use to execute the code cleanup.

ArneMaes0 commented on protocol.xml:
```diff
@@ -5960,6 +6041,14 @@ Are you sure you want to continue.]]>
+			<Content>
+				<Action>16</Action>
+			</Content>
+		</Group>
```
If you reuse the Poll Managers Cleanup part, you no longer need these group, Action and Timer

ArneMaes0 commented on QAction_9000049/QAction_9000049.cs:
```diff
@@ -0,0 +1,45 @@
+	/// The QAction entry point.
+	/// </summary>
+	/// <param name="protocol">Link with SLProtocol process.</param>
+	public static void Run(SLProtocol protocol)
+	{
+		try
+		{
+			var interval = TimeSpan.FromSeconds(Convert.ToDouble(protocol.GetParameter(Parameter.maxtimeforiacmessages_9000049)));
+			var timeToDelete = DateTime.Now - interval;
+
+			var columns = protocol.GetColumns(Parameter.Iac_messages.tablePid, new uint[] { Parameter.Iac_messages.Idx.iac_messagesguid_9000101, Parameter.Iac_messages.Idx.iac_messagesrequesttime_9000108 });
+			var keys = (object[])columns[0];
+			var times = (object[])columns[1];
+
+			var keysToDelete = new List<string>();
+			for (int i = 0; i < keys.Length; i++)
+			{
+				if (DateTime.FromOADate(Convert.ToDouble(times[i])) < timeToDelete)
+				{
+					keysToDelete.Add(Convert.ToString(keys[i]));
+				}
+			}
+
+			protocol.DeleteRows(Parameter.Iac_messages.tablePid, keysToDelete);
+		}
+		catch (Exception ex)
+		{
+			protocol.Log($"QA{protocol.QActionID}|{protocol.GetTriggerParameter()}|Run|Exception thrown:{Environment.NewLine}{ex}", LogType.Error, LogLevel.NoLogging);
+		}
+	}
```
You can also use the 'Skyline.DataMiner.Utils.TableCleanup' nuget package to do these kind of cleanup. We can then also add a max row count just like the cleanup example:
[SLC-C-Example_Table-Cleanup](https://github.com/SkylineCommunications/SLC-C-Example_Table-Cleanup)

This code could then also move to QAction1/Tables/TableCleanup.cs, which is triggered by the Poll Manager and configurable by the end user.

ArneMaes0 commented on QAction_1/Tables/IacMessagesTable.cs:
```diff
@@ -49,6 +50,8 @@ public IAC_MessagesTableRow(params object[] row)
 
 		public string Info { get; set; }
 
+		public DateTime RequestTime { get; set; }
```
Here then also rename this to "Received At"

ArneMaes0 committed:
Renamed the RequestTime column to Received At
Removed QAction 9000049 that does the table cleanup and moved it to QAction1
Changed the TableCleanup to use the Skyline.DataMiner.Utils.TableCleanup nuget package
Added option to clean by row count or max age or both
Removed the minute timer as everything should go through the poll manager

ArneMaes0 commented:
@BlakeSLC Do you prefer to check my changes before I merge it in?

ArneMaes0 commented:
Great I'll merge it and release the new version
